rmf_fleet_adapter_python/scripts/rmf_fleet_adapter_python/test_adapter.py

```diff
@@ -482,132 +577,72 @@ def updater_inserter(handle_obj, updater):
-    flaky_dispenser = MockFlakyDispenser(flaky_dispenser_name)
+    dispenser = MockDispenser(dispenser_name)
+    ingestor = MockIngestor(ingestor_name)
+
+    # INIT TASK SUMMARY OBSERVER ==============================================
+    observer = TaskSummaryObserver()
 
     # FINAL PREP ==============================================================
     rclpy_executor = SingleThreadedExecutor()
-    rclpy_executor.add_node(task_node)
     rclpy_executor.add_node(cmd_node)
-    rclpy_executor.add_node(quiet_dispenser)
-    rclpy_executor.add_node(flaky_dispenser)
-
-    last_quiet_state = False
-    last_flaky_state = False
+    rclpy_executor.add_node(dispenser)
+    rclpy_executor.add_node(ingestor)
+    rclpy_executor.add_node(observer)
 
     # GO! =====================================================================
     adapter.start()
 
     print("# SENDING NEW REQUEST ############################################")
-    request = adpt.type.CPPDeliveryMsg("test_delivery",
+    test_name = 'test_delivery'
+    request = adpt.type.CPPDeliveryMsg(test_name,
                                        pickup_name,
-                                       quiet_dispenser_name,
+                                       dispenser_name,
                                        dropoff_name,
-                                       flaky_dispenser_name)
-    quiet_dispenser.reset()
-    flaky_dispenser.reset()
+                                       ingestor_name)
+    dispenser.reset()
+    ingestor.reset()
+    observer.add_task(test_name)
     adapter.request_delivery(request)
     rclpy_executor.spin_once(1)
 
     for i in range(1000):
-        if quiet_dispenser.success_flag != last_quiet_state:
-            last_quiet_state = quiet_dispenser.success_flag
-            print("== QUIET DISPENSER FLIPPED SUCCESS STATE ==",
-                  last_quiet_state)
-
-        if flaky_dispenser.success_flag != last_flaky_state:
-            last_flaky_state = flaky_dispenser.success_flag
-            print("== FLAKY DISPENSER FLIPPED SUCCESS STATE ==",
-                  last_flaky_state)
-
-        if quiet_dispenser.success_flag and flaky_dispenser.success_flag:
-            rclpy_executor.spin_once(1)
+        if observer.all_tasks_complete():
+            print("All Tasks Complete.")
             break
-
         rclpy_executor.spin_once(1)
-        time.sleep(0.2)
+        # time.sleep(0.2)
 
+    results = observer.count_successful_tasks()
     print("\n== DEBUG TASK REPORT ==")
     print("Visited waypoints:", robot_cmd.visited_waypoints)
-    print("At least one incomplete:", at_least_one_incomplete)
-    print("Completed:", completed)
-    print()
+    print(f"Sucessful Tasks: {results[0]} / {results[1]}")
+
+    assert results[0] == results[1]  # All tasks complete
 
-    assert len(robot_cmd.visited_waypoints) == 6
-    assert all([x in robot_cmd.visited_waypoints for x in [0, 5, 6, 7, 8, 10]])
-    assert robot_cmd.visited_waypoints[0] == 2
```

Comment:
Regarding the aforementioned missing starting waypoint issue, agree, this seems to be related to the source `rmf_core`. I will also need to check about this. But for now this won't be an immediate problem as long as the remaining "visited_waypoints" are correct. Appreciate your help in enhancing this!

Reply:
closing for now